File: mobile/api/goals.ts
/**
 * Typed clients for the FTY-106 goal endpoint and FTY-095 target
 * override/reset endpoints.
 *
 * Endpoints consumed:
 *   POST /api/users/{user_id}/goal               — create/replace active goal (FTY-106)
 *   GET  /api/users/{user_id}/target             — target read-model (FTY-095)
 *   PUT  /api/users/{user_id}/target/override    — set override (FTY-095)
 *   POST /api/users/{user_id}/target/override/reset — reset override (FTY-095)
 *
 * Privacy: calorie, macro, and body-metric values are sensitive derived body
 * data. Errors carry only the HTTP status and the action, never target numbers.
 */

import {
  ApiError,
  authHeaders,
  request,
  userScopedUrl,
} from '@/api/client';
import type { ApiSession } from '@/api/client';
import type { TargetReadModel } from '@/api/dailySummary';

/** Direction of a weight goal. */
export type GoalDirection = 'loss' | 'gain' | 'maintain';

/** Evidence-based pace preset for loss/gain goals. */
export type PacePreset = 'gentle' | 'steady' | 'faster';

/** One independently overridable target key. */
export type OverridableTargetKey = 'calories' | 'protein' | 'carbs' | 'fat';

/**
 * The caller's current active goal, summarised (`GET /goal`, FTY-189/FTY-190).
 *
 * The single authoritative source of a returning user's goal: neither the
 * daily-summary nor the target read-model carries direction or pace, and the
 * in-memory cross-screen seam only knows a goal created/edited *this* session.
 * Both fields are recovered server-side from the persisted trajectory — direction
 * from its sign, pace as the exact inverse of the band it was derived from — so a
 * cold launch reads the real goal. `pace` is `null` for a maintain goal or a
 * legacy goal off the band grid. Trends reads `direction` to colour the weight
 * delta by progress toward the goal; Settings reads both to summarise the Goal row
 * as direction + pace.
 */
export interface ActiveGoal {
  readonly direction: GoalDirection;
  readonly pace: PacePreset | null;
}

/** Persisted goal representation returned by the goal endpoint. */
interface GoalDTO {
  readonly id: string;
  readonly user_id: string;
  readonly start_weight_kg: number;
  readonly start_date: string;
  readonly target_weight_kg: number;
  readonly target_date: string;
  readonly is_active: boolean;
}

/** Revealed target info returned alongside the created goal (FTY-106). */
interface RevealedTarget {
  readonly calories: number;
  readonly rmr_kcal: number;
  readonly tdee_kcal: number;
  readonly direction: GoalDirection;
  readonly clamped: boolean;
}

/** Combined goal + target reveal returned by POST /goal (FTY-106). */
export interface GoalTargetResponse {
  readonly goal: GoalDTO;
  readonly target: RevealedTarget;
  readonly provenance: {
    readonly source: 'derived' | 'user';
    readonly basis: 'goal_and_metrics';
  };
  readonly clamp: {
    readonly clamped: boolean;
    readonly reason: string | null;
  };
}

/** Request body for creating/replacing the active goal. */
export interface GoalTargetRequest {
  /** lose / maintain / gain */
  readonly direction: GoalDirection;
  /** Required for loss/gain; ignored for maintain. */
  readonly pace?: PacePreset;
}

/** Payload for setting manual target overrides (FTY-095). */
export interface TargetOverridePayload {
  readonly calorie_target_kcal?: number;
  readonly protein_target_g?: number;
  readonly carbs_target_g?: number;
  readonly fat_target_g?: number;
}

/** Authenticated session needed to reach these endpoints. */
export type GoalsSession = ApiSession;

/** Raised when a goal or target endpoint returns a non-2xx status. */
export class GoalsApiError extends ApiError {
  constructor(status: number, message: string) {
    super(status, message);
    this.name = 'GoalsApiError';
  }
}

function goalsError(status: number, action: string): GoalsApiError {
  const message =
    status === 401
      ? 'Your session has expired. Sign in again to manage your goal.'
      : status === 404
        ? 'No active goal or target found.'
        : status === 409
          ? 'Complete your profile before setting a goal.'
          : status === 422
            ? 'That goal or override value is not valid.'
            : `Could not ${action} (status ${status}).`;
  return new GoalsApiError(status, message);
}

/**
 * Create or replace the caller's active goal and get the computed target reveal.
 * Returns the new goal, the revealed calorie target, and clamp info. Call
 * `getTarget` afterward to read the full macro target read-model.
 */
export async function createGoal(
  session: GoalsSession,
  payload: GoalTargetRequest,
  fetchImpl: typeof fetch = fetch,
): Promise<GoalTargetResponse> {
  return request<GoalTargetResponse>(
    userScopedUrl(session, 'goal'),
    {
      method: 'POST',
      headers: authHeaders(session),
      body: JSON.stringify(payload),
      action: 'create your goal',
      onError: goalsError,
      fetchImpl,
    },
  );
}

/**
 * Read the caller's current active goal (direction + recovered pace), or `null`
 * when there is none (FTY-189/FTY-190). A `404` — the fail-closed "no active
 * goal" (or cross-user) response — is mapped to `null` rather than thrown: an
 * absent goal is an expected state (a user who has not set one), and the caller
 * treats `null` as an unknown/neutral goal, never a guessed default. `pace` is
 * `null` for a maintain goal or a legacy goal off the band grid. Any other status
 * still throws a `GoalsApiError`.
 */
export async function getActiveGoal(
  session: GoalsSession,
  fetchImpl: typeof fetch = fetch,
): Promise<ActiveGoal | null> {
  try {
    const model = await request<ActiveGoal>(
      userScopedUrl(session, 'goal'),
      {
        method: 'GET',
        headers: authHeaders(session),
        action: 'load your goal',
        onError: goalsError,
        fetchImpl,
      },
    );
    return { direction: model.direction, pace: model.pace ?? null };
  } catch (error) {
    if (error instanceof GoalsApiError && error.status === 404) {
      return null;
    }
    throw error;
  }
}

/**
 * Read just the direction of the caller's active goal, or `null` when there is
 * none. A thin projection of {@link getActiveGoal} for callers (Trends) that only
 * colour by direction and never surface pace.
 */
export async function getActiveGoalDirection(
  session: GoalsSession,
  fetchImpl: typeof fetch = fetch,
): Promise<GoalDirection | null> {
  const goal = await getActiveGoal(session, fetchImpl);
  return goal ? goal.direction : null;
}

/**
 * Fetch the caller's active-goal target for today: calorie + macro targets
 * each with their effective value, derived value, and `derived | user` source.
 * Returns null-shaped read-model on 404 (no active target for today).
 */
export async function getTarget(
  session: GoalsSession,
  fetchImpl: typeof fetch = fetch,
): Promise<TargetReadModel> {
  return request<TargetReadModel>(
    userScopedUrl(session, 'target'),
    {
      method: 'GET',
      headers: authHeaders(session),
      action: 'load your target',
      onError: goalsError,
      fetchImpl,
    },
  );
}

/**
 * Set one or more manual overrides on the caller's target for today.
 * Returns the updated target read-model with `source: 'user'` for each
 * overridden component. An out-of-band value is rejected with `GoalsApiError`
 * status 422 — nothing is persisted.
 */
export async function setTargetOverride(
  session: GoalsSession,
  payload: TargetOverridePayload,
  fetchImpl: typeof fetch = fetch,
): Promise<TargetReadModel> {
  return request<TargetReadModel>(
    userScopedUrl(session, 'target/override'),
    {
      method: 'PUT',
      headers: authHeaders(session),
      body: JSON.stringify(payload),
      action: 'save your target override',
      onError: goalsError,
      fetchImpl,
    },
  );
}

/**
 * Reset one or more manual overrides on the caller's target for today back to
 * the derived value. Pass `targets` to reset specific overrides; omit to reset
 * all in-force overrides. Idempotent.
 */
export async function resetTargetOverride(
  session: GoalsSession,
  targets?: OverridableTargetKey[],
  fetchImpl: typeof fetch = fetch,
): Promise<TargetReadModel> {
  const body = targets ? JSON.stringify({ targets }) : JSON.stringify({});
  return request<TargetReadModel>(
    userScopedUrl(session, 'target/override/reset'),
    {
      method: 'POST',
      headers: authHeaders(session),
      body,
      action: 'reset your target',
      onError: goalsError,
      fetchImpl,
    },
  );
}

File: backend/app/routers/goals.py
"""Goal + target-reveal route with object-level authorization (FTY-106).

A single authenticated, owner-scoped write that turns onboarding's inputs — a
**direction** and an evidence-based **pace preset** — into a persisted active goal
and the **computed daily calorie target returned with its provenance**:

- ``POST /api/users/{user_id}/goal`` — create/replace the active goal and reveal
  its target (``goal`` + ``target`` + ``provenance`` + ``clamp``).

The ``{user_id}`` path is explicit so object-level ownership is checked on every
access; a cross-user request and a missing/unowned goal are indistinguishable and
both fail closed as ``404`` (no existence oracle). An incomplete profile is refused
``409`` (complete the profile first), and a pace not offered for the direction is
refused ``422``. Weight, RMR, TDEE, and the target are sensitive derived body data
and are never logged.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.db import get_session
from app.deps import CurrentUser
from app.schemas.goals import (
    ActiveGoalSummary,
    GoalTargetRequest,
    GoalTargetResponse,
)
from app.services import goals as goals_service
from app.services.goals import InvalidPace
from app.services.targets import GoalForbidden, IncompleteProfileError

router = APIRouter(prefix="/api/users", tags=["goals"])

_NOT_FOUND = HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="goal not found")


@router.post(
    "/{user_id}/goal",
    response_model=GoalTargetResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_goal(
    user_id: uuid.UUID,
    payload: GoalTargetRequest,
    current_user: CurrentUser,
    session: Annotated[Session, Depends(get_session)],
) -> GoalTargetResponse:
    """Create/replace the caller's active goal and reveal its computed target.

    The new goal deactivates any prior active goal (one active goal per user) and
    today's ``daily_targets`` row is persisted as a side effect, so the reveal — and
    a subsequent ``GET daily-summary`` for today — shows a real number. Fails closed
    ``404`` on cross-user access; ``409`` when the profile is incomplete; ``422`` when
    the pace is not offered for the direction (or the request body is malformed).
    """

    try:
        goal, target = goals_service.create_goal_with_target(
            session, user_id, current_user, payload
        )
    except GoalForbidden as exc:
        raise _NOT_FOUND from exc
    except InvalidPace as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)
        ) from exc
    except IncompleteProfileError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="profile must be completed before a target can be computed",
        ) from exc
    return goals_service.build_goal_target_response(goal, target, payload.direction)


@router.get("/{user_id}/goal", response_model=ActiveGoalSummary)
def read_active_goal_summary(
    user_id: uuid.UUID,
    current_user: CurrentUser,
    session: Annotated[Session, Depends(get_session)],
) -> ActiveGoalSummary:
    """Read the direction and pace summary of the caller's active goal.

    Trends colours the weight delta by progress toward the goal, so it needs the
    active goal's direction for a returning user after a cold launch — the only
    authoritative source, since no read-model otherwise carries it. Settings also
    needs the recovered pace to summarize the collapsed Goal row. Fails closed
    ``404`` on cross-user access *and* when the caller simply has no active goal:
    the two are indistinguishable (no existence oracle). No weight/target number
    is exposed or logged.
    """

    try:
        goal = goals_service.read_active_goal(session, user_id, current_user)
    except GoalForbidden as exc:
        raise _NOT_FOUND from exc
    if goal is None:
        raise _NOT_FOUND
    return ActiveGoalSummary(
        direction=goals_service.direction_of(goal),
        pace=goals_service.pace_of(goal),
    )
